library/src/G4CMPEmpiricalLindhardNIEL.cc

```diff
@@ -0,0 +1,115 @@
+// energy in native G4 energy units.
+//
+// The parameter k is a function of energy
+
+
+G4CMPEmpiricalLindhardNIEL::G4CMPEmpiricalLindhardNIEL() 
+    : Empiricalklow(DBL_MIN), Empiricalkhigh(DBL_MIN), 
+      EmpiricalElow(DBL_MIN), EmpiricalEhigh(DBL_MIN), 
+      EmpiricalkFixed(DBL_MIN), EmpiricalEnergyDependentK(true) {
+
+    G4CMPConfigManager* config = G4CMPConfigManager::Instance();
+
+    // Fetch values from G4CMPConfigManager or set defaults
+    Empiricalklow = (config->GetEmpiricalklow() != DBL_MIN) ? config->GetEmpiricalklow() : 0.040;
+    Empiricalkhigh = (config->GetEmpiricalkhigh() != DBL_MIN) ? config->GetEmpiricalkhigh() : 0.142;
+    EmpiricalElow = (config->GetEmpiricalElow() != DBL_MIN) ? config->GetEmpiricalElow() : 0.39 * keV;
+    EmpiricalEhigh = (config->GetEmpiricalEhigh() != DBL_MIN) ? config->GetEmpiricalEhigh() : 7.0 * keV;
+    EmpiricalkFixed = (config->GetEmpiricalkFixed() != DBL_MIN) ? config->GetEmpiricalkFixed() : 0.158;
+    EmpiricalEnergyDependentK = config->GetEmpiricalEnergyDependentK();
+}
+
+G4double G4CMPEmpiricalLindhardNIEL::
+PartitionNIEL(G4double energy, const G4Material *material, G4double Zin,
+		G4double Ain) const {
+
+  if (!material) {
+    G4Exception("G4CMPEmpiricalLindhardNIEL", "G4CMP1000", FatalErrorInArgument,
+		  "No material passed to partition function");
+    return 1.;		// Won't get here after FATAL error
+  }
+    
+  if (firstCall) {
+        	G4Exception("G4CMPEmpiricalLindhardNIEL", "G4CMP1001", JustWarning,
+		    "Lindhard ionization yield model where the parameter k depends\n"
+            "on the recoil energy is selected. Default settings are for Ge.\n"
+            "If klow, khigh, Elow, Ehigh are not passed, default values from\n"
+            "Phys. Rev. D 105, 122002 will be used.");
+            firstCall = false;
+  }
+    
+  if (EmpiricalEnergyDependentK && ((energy < EmpiricalElow) || (energy > EmpiricalEhigh))) {
+      if (firstCall_E) {
+          G4Exception("G4CMPEmpiricalLindhardNIEL", "G4CMP1002", JustWarning,
+		    "Energy is out of bounds. Lewin-Smith model is used.");
+         firstCall_E = false;
+      }
+      return G4CMPLewinSmithNIEL::PartitionNIEL(energy, material);
+  }
+
+  if (EmpiricalEnergyDependentK && (EmpiricalElow == EmpiricalEhigh))) {
+      
+      G4Exception("G4CMPEmpiricalLindhardNIEL", "G4CMP1003", JustWarning,
+          "Elow and Ehigh are equal. Defaulting to Lewin-Smith model.");
+      
+      return G4CMPLewinSmithNIEL::PartitionNIEL(energy, material);
+  }
+
+  const G4double Z=GetEffectiveZ(material), A=GetEffectiveA(material)/(g/mole);
+  G4Pow* g4pow = G4Pow::GetInstance();		// Fast, tabulated x^(2/3) etc.
+  G4double z23 = g4pow->Z23(Z);
+
+  G4double epsilon = 11.5/keV * energy / (Z*z23*z23);	// * Z^(-7/3)
+  G4double k = EmpiricalEnergyDependentK ? Empiricalklow + ((Empiricalkhigh - Empiricalklow) / (EmpiricalEhigh - EmpiricalElow)) * (energy - EmpiricalElow) : EmpiricalkFixed;
```

Comment:
Move this line to above all the Z,A and G4pow stuff, so that the latter is closer to where it's used.

I'd prefer to use an if block here to make the code more readable.
```
G4double k = EmpiricalkFixed;
if (EmpiricalEnergyDependentK) {
  G4double dk = Empiricalkhigh - Empiricalklow;
  G4double dE = EmpiricalEhigh - EmpiricalElow;
  k = klow + (energy-EmpricalElow)*dk/dE;
}
```

You could even make dk and dE, or the ratio, a data member and calculate it just once!